### PR TITLE
birdwatcher: init at 2.2.4, alice-lg: init at 6.0.0, nixos/birdwatcher: init, nixos/alice-lg: init

### PR DESCRIPTION
**Pull request by Janik-Haag**
###### Description of changes

Adding packages and modules for:
- https://github.com/alice-lg/alice-lg
- https://github.com/alice-lg/birdwatcher

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 1]
```diff
@@ -77,6 +77,10 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - [photoprism](https://photoprism.app/), a AI-Powered Photos App for the Decentralized Web. Available as [services.photoprism](options.html#opt-services.photoprism.enable).
 
+- [alice-lg](github.com/alice-lg/alice-lg), a looking-glass for bpg sessions. Available as [services.alice-lg](#opt-services.alice-lg.enable).
```
Comment (RaitoBezarius) [inline nixos/doc/manual/release-notes/rl-2305.section.md:80]:
```suggestion
- [alice-lg](github.com/alice-lg/alice-lg), a looking-glass for BGP sessions. Available as [services.alice-lg](#opt-services.alice-lg.enable).
```

File: nixos/modules/services/networking/alice-lg.nix [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        defaultText = literalExpression "pkgs.alice-lg";
+        description = lib.mdDoc "The Alice Looking Glass package to use.";
+      };
+      enable = mkEnableOption (lib.mdDoc "Alice Looking Glass");
+
+      settings = mkOption {
+        type = settingsFormat.type;
+        default = { };
+        description = lib.mdDoc ''
+          System-wide configuration for GameMode (/etc/gamemode.ini).
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/alice-lg.nix:24]: This seems irrelevant?

File: nixos/modules/services/networking/alice-lg.nix [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+with lib;
+
+let
+  cfg = config.services.alice-lg;
+  settingsFormat = pkgs.formats.ini { };
+in
+{
+  options = {
+    services.alice-lg = {
+      package = mkOption {
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/alice-lg.nix:12]: Nit (and for the next time): we have `mkPackageOptionMD` IIRC

File: nixos/modules/services/networking/alice-lg.nix [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      settings = mkOption {
+        type = settingsFormat.type;
+        default = { };
+        description = lib.mdDoc ''
+          System-wide configuration for GameMode (/etc/gamemode.ini).
+          alice-lg configuration, for configuration options see the example on [github](https://github.com/alice-lg/alice-lg/blob/main/etc/alice-lg/alice.example.conf)
+        '';
+        example = literalExpression ''
+          {
+
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/alice-lg.nix:29]:
```suggestion
```

File: nixos/modules/services/networking/alice-lg.nix [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        wants = [ "network.target" ];
+        after = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        description = "Alice Looking Glass";
+        serviceConfig = {
+          DynamicUser = true;
+          Type = "simple";
+          Restart = "on-failure";
+          RestartSec = 15;
+          #ExecStart = "${cfg.package}/bin/alice-lg -db-init";
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/alice-lg.nix:72]:
```suggestion
```

File: nixos/modules/services/networking/birdwatcher.nix [hunk 6]
```diff
@@ -0,0 +1,131 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.birdwatcher;
+in
+{
+  meta.buildDocsInSandbox = false;
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/birdwatcher.nix:9]: Why?

File: nixos/modules/services/networking/birdwatcher.nix [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+      flags = mkOption {
+        default = [ ];
+        type = types.listOf types.str;
+        example = [ "-worker-pool-size 16" "-6" ];
+        description = lib.mdDoc ''
+          Flags to append to the program call
+        '';
+      };
+
+      settings = mkOption {
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/birdwatcher.nix:29]: This could be cheaply modified into RFC42 via a (INI?) settings generator I suppose?

File: pkgs/servers/alice-lg/default.nix [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  src = fetchFromGitHub {
+    owner = "alice-lg";
+    repo = "alice-lg";
+    rev = version;
+    hash = "sha256-BdhbHAFqyQc8UbVm6eakbVmLS5QgXhr06oxoc6vYtsM=";
+  };
+
+  vendorSha256 = "sha256-SNF46uUTRCaa9qeGCfkHBjyo4BWOlpRaTDq+Uha08y8=";
+
+  ui = stdenv.mkDerivation {
```
Comment (RaitoBezarius) [inline pkgs/servers/alice-lg/default.nix:25]: This should be added to the passthru `passthru.ui = ui;`

File: pkgs/servers/alice-lg/default.nix [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+      yarnLock = src + "/ui/yarn.lock";
+      hash = "sha256-NeK9IM8E2IH09SVH9lMlV3taCmqwlroo4xzmv4Q01jI=";
+    };
+
+    nativeBuildInputs = [ nodejs yarn git ];
+    configurePhase = ''
+      runHook preConfigure
+
+      # Yarn and bundler wants a real home directory to write cache, config, etc to
+      export HOME=$NIX_BUILD_TOP/fake_home
```
Comment (RaitoBezarius) [inline pkgs/servers/alice-lg/default.nix:40]: nit: We usually do `$(mktemp -d)` for HOME folders, YMMV.

File: pkgs/servers/alice-lg/default.nix [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+      ./node_modules/.bin/react-scripts build
+
+      runHook postBuild
+    '';
+
+    installPhase = ''
+      runHook preInstall
+
+      ls -la
```
Comment (RaitoBezarius) [inline pkgs/servers/alice-lg/default.nix:64]:
```suggestion
```

File: pkgs/servers/alice-lg/default.nix [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+      mv build $out
+
+      runHook postInstall
+    '';
+  };
+
+  preBuild = ''
+    cp -R ${ui}/ ui/build/
```
Comment (RaitoBezarius) [inline pkgs/servers/alice-lg/default.nix:72]: does this have to be a `cp` rather than a `lndir`/`ln`?

File: pkgs/servers/alice-lg/default.nix [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  '';
+
+  subPackages = [ "cmd/alice-lg" ];
+  doCheck = false;
+
+  meta = with lib; {
+    homepage = "https://github.com/alice-lg/alice-lg";
+    description = "A looking-glass for bpg sessions";
```
Comment (RaitoBezarius) [inline pkgs/servers/alice-lg/default.nix:80]:
```suggestion
    description = "A looking-glass for BGP sessions";
```

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Some minor cleanup but LGTM, I think, it would be awesome to add a NixOS test to detect early breakages.

File: nixos/modules/services/networking/alice-lg.nix (hunk 2 above)
Comment (Janik-Haag) [reply nixos/modules/services/networking/alice-lg.nix]: Yes, I removed it, it was from the module I initially copied.

File: nixos/modules/services/networking/alice-lg.nix (hunk 3 above)
Comment (Janik-Haag) [reply nixos/modules/services/networking/alice-lg.nix]: Changed it anyways, shouldn't really matter

File: nixos/modules/services/networking/birdwatcher.nix (hunk 7 above)
Comment (Janik-Haag) [reply nixos/modules/services/networking/birdwatcher.nix:29]: I tried that but it broke down when trying to build something like that:
```
          [server]
          allow_from = []
          allow_uncached = false
          modules_enabled = ["status",
                             "protocols",
                             "protocols_bgp",
                             "protocols_short",
                             "routes_protocol",
                             "routes_peer",
                             "routes_table",
                             "routes_table_filtered",
                             "routes_table_peer",
                             "routes_filtered",
                             "routes_prefixed",
                             "routes_noexport",
                             "routes_pipe_filtered_count",
                             "routes_pipe_filtered"
                            ]
```
For the alice-lg module it worked fine so I used it for that.

File: nixos/modules/services/networking/birdwatcher.nix (hunk 6 above)
Comment (Janik-Haag) [reply nixos/modules/services/networking/birdwatcher.nix:9]: Becuase in the line 64 because the test suggested doing that, before some one in the nixpkgs matrix room explained to me how I can properly escape `${pkgs.bird}` with `''${pkgs.bird}`. I can probably remove it now and it should still build.

File: pkgs/servers/alice-lg/default.nix (hunk 9 above)
Comment (Janik-Haag) [reply pkgs/servers/alice-lg/default.nix:40]: I used the gitlab definition as inspiration, that does it like this. I can change it if you want me to.
```
pkgs/applications/version-management/gitlab/default.nix
94:      export HOME=$NIX_BUILD_TOP/fake_home
```

File: pkgs/servers/alice-lg/default.nix (hunk 11 above)
Comment (Janik-Haag) [reply pkgs/servers/alice-lg/default.nix:72]: lndir doesn't exist on my system and `ln -s -d ${ui} ui/build` for a symbolic link of the directory causes the go build to throw an error: `ui/assets.go:12:12: pattern build/*: cannot embed file build/asset-manifest.json: in non-directory build`

File: pkgs/servers/alice-lg/default.nix (hunk 8 above)
Comment (Janik-Haag) [reply pkgs/servers/alice-lg/default.nix:25]: When trying to do that nix-build complains about it not finding the ui attribute:
```
nix-build -A alice-lg
error: undefined variable 'ui'

       at /home/janik/Projects/nixos/nixpkgs/pkgs/servers/alice-lg/default.nix:71:13:

           70|   preBuild = ''
           71|     cp -R ${ui}/ ui/build/
             |             ^
           72|   '';
```

but I'm quite unfamiliar with passthru and the [manual](https://nixos.org/manual/nixpkgs/stable/#special-variables) isn't to helpful with this.

**Comment by Janik-Haag** [comment]
@RaitoBezarius thanks for the review, to be honest I fell like this was quite sloppy work of mine, gonna try better next time. Should be fine now. I will look into writing tests for NixOS modules in a bit only wrote some for home-manger until now, please just give me a bit more time :)

File: pkgs/servers/alice-lg/default.nix (hunk 8 above)
Comment (RaitoBezarius) [reply pkgs/servers/alice-lg/default.nix:25]: You need to reference `passthru.ui` as you're in a `rec` block, you can reference all the others attributes.
Otherwise, using functional mkDerivation, you can have `finalAttrs.ui` (grep the repo for examples)

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
UI change and good to go for merge.

File: pkgs/servers/alice-lg/default.nix (hunk 8 above)
Comment (Janik-Haag) [reply pkgs/servers/alice-lg/default.nix]: Okay done

**Comment by RaitoBezarius** [comment]
Don't forget tests in a follow-up PR :P

**Comment by Janik-Haag** [comment]
> Don't forget tests in a follow-up PR :P

you can also leave this one open for some more time and I will try learning how to write tests this weekend :)

**Comment by RaitoBezarius** [comment]
> > Don't forget tests in a follow-up PR :P
> 
> you can also leave this one open for some more time and I will try learning how to write tests this weekend :)

Sure, but if for some reason, we miss the branch-off, it might not be part of 23.05 FWIW. :)

**Comment by Janik-Haag** [comment]
I think it's ready to get merged now, writing tests was quite a good idea since I discovered a small mistake I made in the alice-lg module which is now fixed :)

I know the tests are quite basic but should be enough to ensure both services start properly.

**Review by RaitoBezarius — APPROVED** [review]
Awesome :)

**Comment by RaitoBezarius** [comment]
Once CI passes, let's merge
